File: paper/origin/src/main/java/com/github/theword/queqiao/EventProcessor.java
package com.github.theword.queqiao;

import com.github.theword.queqiao.tool.GlobalContext;
import com.github.theword.queqiao.tool.event.PlayerAchievementEvent;
import com.github.theword.queqiao.tool.event.PlayerCommandEvent;
import com.github.theword.queqiao.tool.event.model.achievement.AchievementModel;
import com.github.theword.queqiao.tool.event.model.death.DeathModel;
import io.papermc.paper.event.player.AsyncChatEvent;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.TranslatableComponent;
import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer;
import org.bukkit.advancement.Advancement;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.PlayerAdvancementDoneEvent;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;

import static com.github.theword.queqiao.tool.utils.Tool.isRegisterOrLoginCommand;
import static com.github.theword.queqiao.utils.PaperTool.*;

public class EventProcessor implements Listener {

    /**
     * 监听玩家聊天
     *
     * @param event 玩家聊天事件
     */
    @EventHandler(priority = EventPriority.MONITOR)
    void onPlayerChat(AsyncChatEvent event) {
        if (event.isCancelled() || !GlobalContext.getConfig().getSubscribeEvent().isPlayerChat()) return;
        com.github.theword.queqiao.tool.event.PlayerChatEvent spigotAsyncPlayerChatEvent = new com.github.theword.queqiao.tool.event.PlayerChatEvent(getPaperPlayer(event.getPlayer()), "", getComponentJson(event.originalMessage()), getComponentText(event.message()));
        GlobalContext.sendEvent(spigotAsyncPlayerChatEvent);
    }

    /**
     * 监听玩家死亡事件
     *
     * @param event 玩家死亡事件
     */
    @EventHandler
    void onPlayerDeath(PlayerDeathEvent event) {
        if (!GlobalContext.getConfig().getSubscribeEvent().isPlayerDeath()) return;

        DeathModel deathModel = new DeathModel();
        Component component = event.deathMessage();
        if (component instanceof TranslatableComponent translatableComponent) {
            deathModel.setKey(translatableComponent.key());
            String[] args = translatableComponent.args().stream()
                    .map(arg -> {
                        if (arg instanceof TextComponent) {
                            return ((TextComponent) arg).content();
                        } else {
                            return String.valueOf(arg);
                        }
                    })
                    .toArray(String[]::new);
            deathModel.setArgs(args);
        }

        deathModel.setText(getComponentText(component));

        com.github.theword.queqiao.tool.event.PlayerDeathEvent spigotPlayerDeathEvent = new com.github.theword.queqiao.tool.event.PlayerDeathEvent(getPaperPlayer(event.getEntity()), deathModel);
        GlobalContext.sendEvent(spigotPlayerDeathEvent);
    }

    /**
     * 监听玩家加入事件
     *
     * @param event 玩家加入事件
     */
    @EventHandler
    void onPlayerJoin(PlayerJoinEvent event) {
        if (!GlobalContext.getConfig().getSubscribeEvent().isPlayerJoin()) return;

        com.github.theword.queqiao.tool.event.PlayerJoinEvent spigotPlayerJoinEvent = new com.github.theword.queqiao.tool.event.PlayerJoinEvent(getPaperPlayer(event.getPlayer()));
        GlobalContext.sendEvent(spigotPlayerJoinEvent);
    }

    /**
     * 监听玩家离开事件
     *
     * @param event 玩家离开事件
     */
    @EventHandler
    void onPlayerQuit(PlayerQuitEvent event) {
        if (!GlobalContext.getConfig().getSubscribeEvent().isPlayerQuit()) return;

        com.github.theword.queqiao.tool.event.PlayerQuitEvent spigotPlayerQuitEvent = new com.github.theword.queqiao.tool.event.PlayerQuitEvent(getPaperPlayer(event.getPlayer()));
        GlobalContext.sendEvent(spigotPlayerQuitEvent);
    }

    /**
     * 监听玩家命令
     *
     * @param event 玩家命令事件
     */
    @EventHandler
    void onPlayerCommand(PlayerCommandPreprocessEvent event) {
        if (!GlobalContext.getConfig().getSubscribeEvent().isPlayerCommand()) return;

        String command = isRegisterOrLoginCommand(event.getMessage());

        if (command.isEmpty()) return;

        PlayerCommandEvent spigotPlayerCommandPreprocessEvent = new PlayerCommandEvent(getPaperPlayer(event.getPlayer()), "", command, command);
        GlobalContext.sendEvent(spigotPlayerCommandPreprocessEvent);
    }

    @EventHandler
    void onPlayerAdvancement(PlayerAdvancementDoneEvent event) {
        if (!GlobalContext.getConfig().getSubscribeEvent().isPlayerAdvancement()) return;
        Advancement advancement = event.getAdvancement();
        if (advancement.getDisplay() == null || !advancement.getDisplay().doesAnnounceToChat() || event.message() == null)
            return;

        AchievementModel achievementModel = getPaperAdvancement(advancement);
        achievementModel.setText(getComponentText(event.message()));

        PlayerAchievementEvent spigotPlayerAdvancementDoneEvent = new PlayerAchievementEvent(getPaperPlayer(event.getPlayer()), achievementModel);
        GlobalContext.sendEvent(spigotPlayerAdvancementDoneEvent);
    }
}

File: paper/origin/src/main/kotlin/com/github/theword/queqiao/EventProcessor.kt
package com.github.theword.queqiao

import com.github.theword.queqiao.tool.GlobalContext
import com.github.theword.queqiao.tool.event.PlayerAchievementEvent
import com.github.theword.queqiao.tool.event.PlayerChatEvent
import com.github.theword.queqiao.tool.event.PlayerCommandEvent
import com.github.theword.queqiao.tool.event.PlayerDeathEvent
import com.github.theword.queqiao.tool.event.PlayerJoinEvent
import com.github.theword.queqiao.tool.event.PlayerQuitEvent
import com.github.theword.queqiao.tool.event.model.PlayerModel
import com.github.theword.queqiao.tool.event.model.TranslateModel
import com.github.theword.queqiao.tool.event.model.achievement.AchievementModel
import com.github.theword.queqiao.tool.utils.Tool
import com.github.theword.queqiao.utils.PaperTool.Companion.getComponentJson
import com.github.theword.queqiao.utils.PaperTool.Companion.getComponentText
import com.github.theword.queqiao.utils.PaperTool.Companion.getPaperAdvancement
import com.github.theword.queqiao.utils.PaperTool.Companion.getPaperPlayer
import com.github.theword.queqiao.utils.PaperTool.Companion.parseTranslateModel
import net.kyori.adventure.text.Component
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener


class EventProcessor : Listener {


    /**
     * 监听玩家聊天
     *
     * @param event 玩家聊天事件
     */
    @EventHandler(priority = EventPriority.MONITOR)
    fun onPlayerChat(event: io.papermc.paper.event.player.AsyncChatEvent) {
        if (event.isCancelled || !GlobalContext.getConfig().subscribeEvent.isPlayerChat) return
        val spigotAsyncPlayerChatEvent = PlayerChatEvent(
            getPaperPlayer(event.getPlayer()),
            "",
            getComponentJson(event.originalMessage()),
            getComponentText(event.message())
        )
        GlobalContext.sendEvent(spigotAsyncPlayerChatEvent)
    }

    /**
     * 监听玩家死亡事件
     *
     * @param event 玩家死亡事件
     */
    @EventHandler
    fun onPlayerDeath(event: org.bukkit.event.entity.PlayerDeathEvent) {
        if (!GlobalContext.getConfig().subscribeEvent.isPlayerDeath) return

        val component: Component? = event.deathMessage()

        val translateModel: TranslateModel = parseTranslateModel(component)
        val spigotPlayerDeathEvent = PlayerDeathEvent(getPaperPlayer(event.getEntity()), translateModel)
        GlobalContext.sendEvent(spigotPlayerDeathEvent)
    }

    /**
     * 监听玩家加入事件
     *
     * @param event 玩家加入事件
     */
    @EventHandler
    fun onPlayerJoin(event: org.bukkit.event.player.PlayerJoinEvent) {
        if (!GlobalContext.getConfig().subscribeEvent.isPlayerJoin) return

        val spigotPlayerJoinEvent = PlayerJoinEvent(getPaperPlayer(event.player))
        GlobalContext.sendEvent(spigotPlayerJoinEvent)
    }

    /**
     * 监听玩家离开事件
     *
     * @param event 玩家离开事件
     */
    @EventHandler
    fun onPlayerQuit(event: org.bukkit.event.player.PlayerQuitEvent) {
        if (!GlobalContext.getConfig().subscribeEvent.isPlayerQuit) return

        val spigotPlayerQuitEvent = PlayerQuitEvent(getPaperPlayer(event.player))
        GlobalContext.sendEvent(spigotPlayerQuitEvent)
    }

    /**
     * 监听玩家命令
     *
     * @param event 玩家命令事件
     */
    @EventHandler
    fun onPlayerCommand(event: org.bukkit.event.player.PlayerCommandPreprocessEvent) {
        val command: String = Tool.isIgnoredCommand(event.message)

        if (command.isEmpty()) return

        if (!GlobalContext.getConfig().subscribeEvent.isPlayerCommand) return

        val spigotPlayerCommandPreprocessEvent =
            PlayerCommandEvent(getPaperPlayer(event.getPlayer()), "", command, command)
        GlobalContext.sendEvent(spigotPlayerCommandPreprocessEvent)
    }

    @EventHandler
    fun onPlayerAdvancement(event: org.bukkit.event.player.PlayerAdvancementDoneEvent) {
        val advancement = event.advancement
        if (advancement.display == null || !advancement.display!!.doesAnnounceToChat() || event.message() == null) return
        if (!GlobalContext.getConfig().subscribeEvent.isPlayerAdvancement) return
        val player: PlayerModel = getPaperPlayer(event.getPlayer())
        val achievementModel: AchievementModel = getPaperAdvancement(player.nickname, advancement)

        val spigotPlayerAdvancementDoneEvent = PlayerAchievementEvent(player, achievementModel)
        GlobalContext.sendEvent(spigotPlayerAdvancementDoneEvent)
    }

}